### PR TITLE
browse department as directory

### PR DESCRIPTION
**Pull request by tomx4096**
## Status

Ready for review / in progress

## Description of Changes

Related to #527 #474 #500 , provides a directory view of the department that users can quickly Ctrl-F through if they know name or star number. 

Changes proposed in this pull request:

 - Add a departmental view button to browse.html
 - Leads to directory page that listing all officers in a department, with links to officer profiles
 - 

## Notes for Deployment

- Depends on maintaining ranks in RANK_CHOICES in high-to-low order, excluding Not Sure
- Only Commander, Captain, PO, and Not Sure rank included in template right now

## Screenshots (if appropriate)

screenshot from 2018-08-12 03-19-20
screenshot from 2018-08-12 03-19-09
screenshot from 2018-08-12 03-18-44
screenshot from 2018-08-12 03-18-57

## Tests and linting

 - [x] I have rebased my changes on current `develop`

 - [x] pytests pass in the development environment on my local machine

 - [x] `flake8` checks pass

### CODE REVIEW

**Comment by dismantl** [comment]
This is great, thanks! I was able to get it to work with the test data, but when I tried to use Baltimore data the ranks don't match up with what's in `RANK_CHOICES` and so it throws an error. Is there a way this could gracefully handle departments that have different sets of ranks?

**Comment by tomx4096** [comment]
Thanks for the feedback. You're right, it could be hard to manually maintain proper rank names for many departments with slight variations. I pushed another approach that makes it independent of RANK_CHOICES. It makes up a list of ranks found in the data and uses those to demarcate sections in the directory view. The html was also generalized so all ranks found would be displayed, with Not Sure at the end, and a random (blue-ish) background color. 

The disadvantage of this is that they aren't guaranteed to be ordered from high to low rank in the directory view. 

Let me know if it fixes your error, I can also try your dataset if you like, just let me know where to find it.

File: OpenOversight/app/main/views.py [hunk 1]
```diff
@@ -305,6 +305,47 @@ def list_officer(department_id, page=1, from_search=False):
...
+            assign_dict[record.officer_id] = record
+        else:
+            if record.star_date > assign_dict[record.officer_id].star_date:
+                assign_dict[record.officer_id] = record
+
+    officer_nodes = {}
+    rank_types = ["NOT SURE"]
+
+    for officer in officers:
+        rec = assign_dict[officer.id]
```
Comment (dismantl) [inline OpenOversight/app/main/views.py]: It's possible there exist officers that don't have assignments, so this should probably be wrapped in a try block:
```
try:
    rec = assign_dict[officer.id]
except KeyError:
    continue
```

**Comment by dismantl** [comment]
Cool, I like that approach and I was able to get it working for Baltimore data. We have like dozens of job titles, some with only 1 or 2 employees, so that made the list pretty long and unwieldy. But the solution to that is probably just to filter out what officers I import (e.g. not forensic scientists).

I wonder if the best way to allow department-specific ranks but not have the disadvantages you listed above, would be to have a per-department list of ranks provided when a department is created. Maybe by uploading an ordered CSV or otherwise storing it in the database? But that may also be a follow-on issue beyond the scope of this PR :)

File: OpenOversight/app/main/views.py [hunk 2]
```diff
@@ -305,6 +305,47 @@ def list_officer(department_id, page=1, from_search=False):
...
+        if len(rec.rank) <= 1:
+            rec.rank = "NOT SURE"
+        safe_rank = rec.rank.upper()
+        if safe_rank not in rank_types:
+            # keep NOT SURE at the end so it will display last on page
+            rank_types.insert(0, safe_rank)
+        if safe_rank not in officer_nodes.keys():
+            officer_nodes[safe_rank] = []
+        else:
+            officer_nodes[safe_rank].append({'id': officer.id,
```
Comment (dismantl) [inline OpenOversight/app/main/views.py]: This shouldn't be in an `else` block or it won't add the first officer of a rank it finds.

File: OpenOversight/app/main/views.py (hunk 1 above)
Comment (tomx4096) [reply OpenOversight/app/main/views.py]: thanks, i hadn't considered this case. fixed in last commit

File: OpenOversight/app/main/views.py (hunk 2 above)
Comment (tomx4096) [reply OpenOversight/app/main/views.py]: ah ha! thanks for catching this, fixed in last commit

**Comment by redshiftzero** [comment]
> I wonder if the best way to allow department-specific ranks but not have the disadvantages you listed above, would be to have a per-department list of ranks provided when a department is created. Maybe by uploading an ordered CSV or otherwise storing it in the database?

:+1: to this, storing the ranks in the database is the right way to go (fine to do in a separate issue/PR imo, see #326)

File: OpenOversight/app/main/views.py [hunk 3]
```diff
@@ -305,6 +305,50 @@ def list_officer(department_id, page=1, from_search=False):
...
 
 
+@main.route('/department/<int:department_id>/chart')
+def dept_org_chart(department_id):
+    department = Department.query.filter_by(id=department_id).first()
+    if not department:
+        abort(404)
+
+    officers = Officer.query.filter(Officer.department_id == department_id).all()
+    # map each officer to their most recent assignment
```
[the remark this answers is not in the document]
Comment (tomx4096) [reply OpenOversight/app/main/views.py]: at first i tried using this relationship, but I found that accessing assignments on each Officer was very slow (sometimes timing out the browser on the test data). i found it much faster to query all Assignments at once and then associate them to officers in the loop. additionally, i wasn't sure that the assignments would be guaranteed to be in chronological order in the list

**Pull request closed without merging**